### DISCUSSION TITLE
Can I extend a struct ?

### DISCUSSION THREAD

**Discussion by gwenn** [bytedeco/javacpp#586] [Q&A] [answered]
Hello,
For this [project](https://github.com/gwenn/sqlite-jna/tree/javacpp), I would like to implement [virtual table](https://sqlite.org/vtab.html#implementation) in Java.
To do so, you need to "extend" C struct like [here](https://github.com/sqlite/sqlite/blob/master/ext/misc/carray.c#L102-L108):
```C
struct my_cursor {
  sqlite3_vtab_cursor base;  /* Base class - must be first */
  // at least one other field here
};
```
Do you know if I can do this with `javacpp` ?
Do I need to write some C code ?
Thanks.

**Comment by saudet** [comment, marked as answer]
It sounds like we have to create a new type, and if that is the case, we need to provide an additional header file to JavaCPP for that, yes.

If you would like to be able to do that kind of thing from Java, JavaCPP already implements a mechanism for virtual methods in C++:
https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#dealing-with-abstract-classes-and-virtual-methods
So that means if you can figure out a way to map that to a standard C++ abstract class, JavaCPP can do the rest.

**Reply by gwenn** [reply]
Thanks for your reply.
So I cannot rewrite my own "jarray" virtual table in pure java / without writing some C/C++ code ?
Like in [rust](https://github.com/rusqlite/rusqlite/blob/master/src/vtab/array.rs#L126-L135) for example.

**Reply by saudet** [reply]
Rust types are compatible with C, similar to how C++ does it too. Java ones are not, so we would need a mechanism to create new C++ types before mapping them. JavaCPP currently provides ways to do that for the common cases of virtual methods and templates in C++, but not for ad hoc ones like this, no.

**Reply by saudet** [reply]
That said, we can allocate a chunk of native memory and do whatever we want with it, all from Java. That could work, but it wouldn't be pretty and wouldn't feel anything like a struct. So, that makes me think, maybe what you're looking for is [Panama FMA](https://github.com/openjdk/panama-foreign/blob/foreign-jextract/doc/panama_memaccess.md)? See issue #200
/cc @jverne @mcimadamore

They expend a lot of effort in supporting all these quirks of the C language, but inside the Java language, without using any C/C++/Rust compiler like GCC, LLVM, MSVC, etc. Unfortunately, they are doing so at the expense of supporting C++ features, but I think anything related to SQLite is 100% C so it should be safe for that particular specific use case, and they already have some instructions for it at http://hg.openjdk.java.net/panama/dev/raw-file/foreign/doc/panama_foreign.html

**Comment by gwenn** [comment]
Thanks fo all these details.
I will try the standard C++ abstract class.

**Reply by gwenn** [reply]
Hello,
I've tried the C++ abstract classes here:
https://github.com/gwenn/sqlite-jna/commit/10018553668ffdc8f504dedafbe445e5302f01f8
But I should be doing something wrong:
https://github.com/gwenn/sqlite-jna/runs/7378415239?check_suite_focus=true
Could you please advise ?
Thanks.

**Reply by saudet** [reply]
```java
	@Virtual(true) public native int Open(@Cast("sqlite3_vtab_cursor**") PointerPointer ppCursor);
	@Virtual(true) public native int Open(@ByPtrPtr sqlite3_vtab_cursor ppCursor);
```
A concrete C++ class gets generated with those declarations, and these 2 functions end up with the same signature in C++, so that can't work.
